### PR TITLE
Recovery

### PR DESCRIPTION
**Pull request by jdonszelmann**
This PR implementes two features (I'm sorry). First of all, bracket recovery.

* There is now a `SourceFile` type bundling a name and source contents
* There is now a `TokenStream` type which is a newtype around the old `Lexer` type
* There is now a `TokenBuffer` type which eagerly parses a tokenstream into a `VecDeque` allowing parser to be more efficient later
    * While a `TokenBuffer` is constructed, an attempt is made to balance parentheses and brackets
    * This may throw *recoverable* errors
* A Parser take a `TokenBuffer` now.
* There is now an `ErrorContext`. Most functions need to take this in in some way. It stores Errors together with the source code they relate to
* There are methods on `ErrorContext` to add errors to it: either fatal or recoverable.
* These methods can fail themselves, fatal *always* errors in fact. This turns the `ErrorContext` into a `CompilerError` which is a struct that holds which error was the final fatal one, together with all previous recoverable errors.
* debug-printing a `CompilerError` prints all recoverable errors, finishing with the fatal one.
* `CompilerError` is generic over an inner error type `E`. This is so that parse tests can tests for specifically a `ParseError` occuring, not just any generic `CompilerError`. Otherwise, we'd basically only have a single error type which is very annoying in testing exhaustively.
* Any generic `CompilerError` can be turned into a dynamic one, by using `CompilerErrorKind`, an enum with conversions from each of the error types. Multiple levels are possible, take a look at `BracketError`
* There are some extension trait methods on various `Result`s to make working with them easier.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        }
+
+        let (e, source) = &self.fatal;
+        write!(f, "{:?}", e.clone().into().to_miette(source))
+    }
+}
+
+/// Holds the current compiler state of recoverable errors until a fatal error is thrown.
+pub struct ErrorContext<'s> {
+    errors: Vec<(CompilerErrorKind, SourceFile<'s>)>,
```
Comment (bal-e) [inline src/error.rs:160]: We can group compiler errors by their source files.  In fact, until the name resolution / type-checking stage, all processing is happening per-file, so I think it's not a bad idea to have a concept of "current source file" in this type (or in a wrapper around this type).

**Review by tertsdiepraam — COMMENTED** [review]
I'm not sure I fully understand your model for error recovery. What I expected was `Result` on functions for fatal errors and recoverable errors logged to a `Vec`. Then at the end combine both into a single `Result`. But here we have fatal errors that are logged but also returned and both are put in a `CompilerError`? What if `CompilerError` was just an `enum` with a variant `Lex(Vec<LexError>)`, that also making it clear that these errors are per stage. Basically, I think the idea is great, but you maybe climbed the abstraction ladder a bit too much.

File: src/parser/ast.rs [hunk 2]
```diff
@@ -92,3 +92,5 @@ pub enum Item<'s, 'a> {
 pub struct File<'s, 'a> {
     pub items: &'a [Item<'s, 'a>],
 }
+
+pub type Ast<'s, 'a> = File<'s, 'a>;
```
Comment (tertsdiepraam) [inline src/parser/ast.rs:96]: I like this :)

File: src/main.rs [hunk 3]
```diff
@@ -1,26 +1,62 @@
...
-    Ok(())
+
+    let ast = match compile(source_file, &arena) {
+        Err(e) => {
+            eprintln!("{:?}", e);
+            return ExitCode::FAILURE;
+        }
+        Ok(i) => i,
+    };
+    println!("{:?}", ast);
```
Comment (tertsdiepraam) [inline src/main.rs:59]: Isn't this basically the same as returning the error?

File: src/parser/function.rs [hunk 4]
```diff
@@ -1,20 +1,20 @@
...
+use crate::lexer::token::Token;
 use crate::parser::ast::{Function, Parameter};
 use crate::parser::error::ParseResult;
 use crate::parser::span::{Spanned, WithSpan};
 use crate::parser::Parser;
 use bumpalo::collections::Vec;
 
-impl<'s, 'a> Parser<'s, 'a> {
-    fn parse_parameter(&mut self) -> ParseResult<Parameter<'s>> {
+impl<'s, 'a, 'e> Parser<'s, 'a, 'e> {
```
Comment (tertsdiepraam) [inline src/parser/function.rs:8]: Why is `'e` added? It's not used in the diff as far as I can tell, but maybe I'm not looking in the right places.

**Comment by tertsdiepraam** [comment]
@jdonszelmann so we can close this definitely?

**Comment by jdonszelmann** [comment]
Yes, I already made it so the other pr closed this one too but you can do it manually

**Comment by tertsdiepraam** [comment]
Let's keep it clean and remove this then

**Pull request closed without merging**